src/cargo/core/compiler/build_runner/compilation_files.rs

```diff
@@ -649,6 +649,23 @@ fn compute_metadata(
     // with user dependencies.
     unit.is_std.hash(&mut hasher);
 
+    // While we don't hash RUSTFLAGS because it may contain absolute paths that
+    // hurts reproducibility, we track whether a unit's RUSTFLAGS is from host
+    // config, so that we can generate a different metadata hash for runtime
+    // and compile-time units.
+    // This is essentially a hack for fixing rust-lang/cargo#14253.
+    if unit.kind.is_host() && !bcx.gctx.target_applies_to_host().unwrap_or_default() {
+        let host_info = bcx.target_data.info(CompileKind::Host);
+        let target_configs_are_different = unit.rustflags != host_info.rustflags
+            || unit.rustdocflags != host_info.rustdocflags
+            || bcx
+                .target_data
+                .target_config(CompileKind::Host)
+                .links_overrides
+                != unit.links_overrides;
```

Comment:
This feels a bit brittle. If we don't remember to update this with anything else that is different but not in metadata already, we'll see this problem again.

The only thoughts I had are
- Putting this question in the unit or host info to answer
- Hashing `unit.kind.is_host() && !bcx.gctx.target_applies_to_host().unwrap_or_default()`

Note: answering this is more important for long term reliability and stabilization.  This doesn't need to block fixing people who are using the unstable feature